packages/node/src/indexer/SpecVersions.service.ts

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020-2022 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import { Injectable } from '@nestjs/common';
+import { ApiPromise } from '@polkadot/api';
+import { BlockHash, RuntimeVersion } from '@polkadot/types/interfaces';
+
+export type SpecVersionMap = { [blockNumber: number]: number }; // blockNumber => specVersion
+
+export function getSpecVersionBlockNumbers(
+  specVersions: SpecVersionMap,
+): number[] {
+  return Object.keys(specVersions).map((key) => parseInt(key, 10));
+}
+
+@Injectable()
+export class SpecVersionService {
+  private _specVersions: SpecVersionMap;
```

Comment:
When dictionary is used, this is fine since its length is very limited. But when dictionary is not used, this could be too many to be kept in memory, and unnecessary.